File: Pythoncode/53.最大子序和.py
class Solution:
    """
    https://leetcode-cn.com/problems/maximum-subarray/solution/zui-da-zi-xu-he-cshi-xian-si-chong-jie-fa-bao-li-f/
    """
    def maxSubArray(self, nums: List[int]) -> int:
        dp = [nums[0]] # dp[i] = 以nums[i]结尾的最大自序和
        for i in range(1, len(nums)):
            dp.append(max(nums[i], nums[i] + dp[i-1]))
        return max(dp)

File: Cppcode/53.最大子序和.cpp
#include <vector>
#include <algorithm>

using namespace std;

class Solution {
public:
    int maxSubArray(vector<int>& nums) {
        vector<int> dp; // dp[i] = 以nums[i]作为结尾的最大子序和
        for (auto i = 0; i < nums.size(); i++)
        {
            if (i == 0 || dp[i-1] <= 0)
            {
                dp.push_back(nums[i]);
            }
            else
            {
                dp.push_back(nums[i]+dp[i-1]);
            }
        }
        return *max_element(dp.begin(), dp.end());
    }
};